compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/ComposeSceneKeyboardOffsetManager.kt

```diff
@@ -0,0 +1,264 @@
+            keyboardHeight = keyboardHeight(targetFrame),
+            viewBottomIndent = bottomIndent,
+            duration = duration,
+            animationOptions = animationOptions
+        )
+    }
+
+    @OptIn(ExperimentalComposeApi::class)
+    private fun animateKeyboard(
+        previousKeyboardHeight: Double,
+        keyboardHeight: Double,
+        viewBottomIndent: Double,
+        duration: Double,
+        animationOptions: UIViewAnimationOptions
+    ) {
+        // Animate view from 0 to [animationTargetSize] and normalize to animation progress with
+        // range of [0..1] to follow UIKit animation curve values.
+        val animationTargetSize = 1000.0
+        val animationTargetFrame = CGRectMake(0.0, 0.0, 0.0, animationTargetSize)
+        fun getCurrentAnimationProgress(): Double {
+            val layer = animationView.layer.presentationLayer() ?: return 0.0
+            return layer.frame.useContents { size.height / animationTargetSize }
+        }
+
+        fun updateAnimationValues(progress: Double) {
+            val currentHeight = previousKeyboardHeight +
+                (keyboardHeight - previousKeyboardHeight) * progress
+            val currentOverlapHeight = max(0.0, currentHeight - viewBottomIndent)
+
+            val targetBottomOffset = calcFocusedBottomOffsetY(currentOverlapHeight)
+            viewBottomOffset += (targetBottomOffset - viewBottomOffset) * progress
+
+            keyboardOverlapHeightState.value = currentOverlapHeight.dp
+        }
+
+        fun completeAnimation() {
+            animationView.removeFromSuperview()
+            updateAnimationValues(1.0)
+        }
+
+        //attach to root view if needed
+        if (animationView.superview == null) {
+            view.addSubview(animationView)
+        }
+
+        //cancel previous animation
+        animationView.layer.removeAllAnimations()
+        keyboardAnimationListener?.invalidate()
+
+        UIView.performWithoutAnimation {
+            animationView.setFrame(CGRectZero.readValue())
+        }
+
+        //animation listener
+        val keyboardDisplayLink = CADisplayLink.displayLinkWithTarget(
+            target = object : NSObject() {
+                @OptIn(BetaInteropApi::class)
+                @Suppress("unused")
+                @ObjCAction
+                fun animationDidUpdate() {
+                    updateAnimationValues(getCurrentAnimationProgress())
+                }
+            },
+            selector = sel_registerName("animationDidUpdate")
+        )
+        keyboardAnimationListener = keyboardDisplayLink
+        UIView.animateWithDuration(
+            duration = duration,
+            delay = 0.0,
+            options = animationOptions or UIViewAnimationOptionBeginFromCurrentState,
+            animations = {
+                animationView.setFrame(animationTargetFrame)
+            },
+            completion = { isFinished ->
+                keyboardDisplayLink.invalidate()
```

Comment:
Should we also set` keyboardAnimationListener` to nil here? Unlike usual target:selector: APIs, CADisplayLink strongly retains its target. Could that lead to a retain-cycle that GC can't break?

Reply:
Thanks, that's make sense! I added nullification code.